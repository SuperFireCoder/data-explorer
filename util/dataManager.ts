import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { KeycloakInstance } from "../interfaces/Keycloak";
import { IHttpResponseError } from "../interfaces/ErrorResponse"
const ENDPOINTS = {
    DATASET: "/api/dataset/",
    PERMISSION: "/api/permission/",
};
import { PinnedDataset } from "./../interfaces/PinnedDataset";
export class DataManager {
    private readonly axios: AxiosInstance;

    constructor(
        public readonly serverBaseUrl: string,
        private readonly keycloakInstance:  KeycloakInstance | undefined
    ) {
        const { axiosInstance } = this.initNewAxiosInstance();
        this.axios = axiosInstance;
    }

    public getAuthorizationHeaderValue() {
        const token = this.keycloakInstance?.token;

        if (token === undefined) {
            return undefined;
        }

        return `Bearer ${token}`;
    }

    private initNewAxiosInstance() {
        const axiosInstance = axios.create({
            baseURL: this.serverBaseUrl,
        });

        // interceptor to handle the response
        axiosInstance.interceptors.response.use(
        (response) => response,
        (error) => {
            const errorResponse: IHttpResponseError = {
                title: '',
                code: "",
                description: ""
              };
        if (error.response){
            {
                errorResponse.title = error.response.data.title;
                errorResponse.code = error.response.data.code;
                errorResponse.description = error.response.data.description? error.response.data.error.description : undefined;
            } 
            } else {
            // Handle other errors
            errorResponse.title = "Something went wrong with that request.";
            errorResponse.code = "GE-DE-001"
            errorResponse.description = "";
        }

        return Promise.reject(errorResponse);
        });

        const injectAuthHeaderInterceptor =
            axiosInstance.interceptors.request.use((requestConfig) => {
                const authHeader = this.getAuthorizationHeaderValue();

                // If no auth value available, just pass request through
                if (authHeader === undefined) {
                    return requestConfig;
                }

                // Otherwise inject the auth header
                return {
                    ...requestConfig,
                    headers: {
                        ...requestConfig.headers,
                        Authorization: authHeader,
                    },
                };
            });

        return { axiosInstance, injectAuthHeaderInterceptor };
    }

    private getNewAxiosCancellationToken() {
        return axios.CancelToken.source();
    }

    public xhrGet<T>(url: string) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        const axiosPromise = this.axios.get<T>(url, {
            cancelToken: cancellationToken.token,
        });
        const promise = axiosPromise.then((res) => res.data);
        return { promise, cancellationToken, axiosPromise };
    }

    private xhrPost<T>(url: string, data: unknown) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        const axiosPromise = this.axios.post<T>(url, data, {
            cancelToken: cancellationToken.token,
        });
        const promise = axiosPromise.then((res) => res.data);
        return { promise, cancellationToken, axiosPromise };
    }

    private xhrDelete<T>(url: string) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        const axiosPromise = this.axios.delete<T>(url, {
            cancelToken: cancellationToken.token,
        });
        
        const promise = axiosPromise.then((res) => res.data);
        return { promise, cancellationToken, axiosPromise };
    }

    private xhrPatch<T>(url: string, data: unknown) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        const axiosPromise = this.axios.patch<T>(url, data, {
            cancelToken: cancellationToken.token,
        });
        const promise = axiosPromise.then((res) => res.data);
        return { promise, cancellationToken, axiosPromise };
    }

    public getDatasetFileStatus(url: string) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        return this.xhrGet<{ url: string, status: string }>(url);
    }


    public getDatasetTemporaryUrl(uuid: string) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        return this.xhrGet<{ url: string, status: string }>(
            `${ENDPOINTS.DATASET}${uuid}/tempurl`
        );
    }


    public removeDataset(uuid: string) {
        return this.xhrDelete<{}>(
            `${ENDPOINTS.DATASET}${uuid}/delete`
        );
    }


    public getDatasetPermissions(uuid: string) {
        return this.xhrGet<Record<string, string[]>>(
            `${ENDPOINTS.PERMISSION}${uuid}`
        );
    }

    public updateDatasetPermissions(
        uuid: string,
        permissions: Record<string, string[]>
    ) {
        return this.xhrPost<unknown>(`${ENDPOINTS.PERMISSION}${uuid}/update`, {
            permissions,
        });
    }

    public removeDatasetPermissions(uuid: string, userIds: string[]) {
        return this.xhrPost<unknown>(`${ENDPOINTS.PERMISSION}${uuid}/delete`, {
            users: userIds,
        });
    }

    public pinDataset(uuid: string) {
        return this.xhrPost<PinnedDataset []>(
            "/userpinneddatasets/",
            {"uuid": uuid}
        );
    }

    public unPinDataset(uuid: string) {
        const url = `/userpinneddatasets/?uuid=${uuid}`;
        return this.xhrDelete<unknown>(url);
    }

    public getPinnedDataset(permissions: Record<string, string|undefined>) {
        return this.xhrGet<PinnedDataset[]>(
            "/userpinneddatasets/",
        );
    }
}
