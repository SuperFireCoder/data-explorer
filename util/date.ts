const shortMonthStrings = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
];

export function getDDMMMYYYY(date: Date) {
    const day = date.getDate();
    const monthIndex = date.getMonth();
    const year = date.getFullYear();

    return `${day} ${shortMonthStrings[monthIndex]} ${year}`;
}
