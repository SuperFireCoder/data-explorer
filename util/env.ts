import getConfig from "next/config";

const config = getConfig();

export const IS_DEVELOPMENT = config.publicRuntimeConfig.NEXT_PUBLIC_DEPLOYMENT !== "production";

export const IS_BSC = String(config.publicRuntimeConfig.NEXT_PUBLIC_KEYCLOAK_AUTH_URL)
                        .match(/biosecuritycommons/) !== null
                        || config.publicRuntimeConfig.NEXT_PUBLIC_THEME === 'bsc-theme';
                        
export function getDataExplorerBackendServerUrl() {
    return config.publicRuntimeConfig
        .NEXT_PUBLIC_DATA_EXPLORER_BACKEND_SERVER_URL as string;
}

export function getUserManagementBackendServerUrl() {
    return config.publicRuntimeConfig
        .NEXT_PUBLIC_USER_MANAGEMENT_BACKEND_SERVER_URL as string;
}

export function getDataExplorerSubbarImportData() {
    return config.publicRuntimeConfig
        .NEXT_PUBLIC_DATA_EXPLORER_SUBBAR_IMPORT_DATA as string | undefined;
}

export function getKeycloakAuthParameters() {
    return {
        url: config.publicRuntimeConfig.NEXT_PUBLIC_KEYCLOAK_AUTH_URL as string,
        realm: config.publicRuntimeConfig
            .NEXT_PUBLIC_KEYCLOAK_AUTH_REALM as string,
        clientId: config.publicRuntimeConfig
            .NEXT_PUBLIC_KEYCLOAK_AUTH_CLIENT_ID as string,
    };
}

export function getGoogleAnalyticsTrackingId() {
    return config.publicRuntimeConfig
        .NEXT_PUBLIC_GOOGLE_ANALYTICS_TRACKING_ID as string;
}

export function getWorkFlowUrl() {
    return config.publicRuntimeConfig
        .NEXT_PUBLIC_WORKFLOW_URL as string;
}

export function getWorkSpaceUrl() {
    return config.publicRuntimeConfig
        .NEXT_PUBLIC_WORKSPACE_URL as string;
}
