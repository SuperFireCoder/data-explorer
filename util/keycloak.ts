import { useMemo } from "react";
import { useKeycloak } from "@react-keycloak/ssr";
import { KeycloakInstance } from "../interfaces/Keycloak";

export function useKeycloakInfo() {

    const { keycloak, initialized } = useKeycloak<KeycloakInstance>();

    /** Whether a user is signed in and has an active session */
    const userSessionActive = useMemo<Boolean>(
        () => keycloak?.token !== undefined,
        [keycloak?.authenticated, keycloak?.token]
    );

    return  { 
        keycloak,
        initialized,
        userSessionActive
    }
}
