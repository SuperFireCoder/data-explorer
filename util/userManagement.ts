import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { UserInfo } from "../interfaces/UserManagement";

import { KeycloakInstance } from "../interfaces/Keycloak";

const ENDPOINTS = {
    USERS: "/api/users/",
};

export class UserManagement {
    private readonly axios: AxiosInstance;

    constructor(
        public readonly serverBaseUrl: string,
        private readonly keycloakInstance: KeycloakInstance
    ) {
        this.axios = this.initNewAxiosInstance();
    }

    public getAuthorizationHeaderValue() {
        return `Bearer ${this.keycloakInstance.token}`;
    }

    private initNewAxiosInstance() {
        return axios.create({
            baseURL: this.serverBaseUrl,
            headers: {
                Authorization: this.getAuthorizationHeaderValue(),
            },
        });
    }

    private getNewAxiosCancellationToken() {
        return axios.CancelToken.source();
    }

    private xhrGet<T>(url: string, config?: AxiosRequestConfig) {
        const cancellationToken = this.getNewAxiosCancellationToken();
        const axiosPromise = this.axios.get<T>(url, {
            cancelToken: cancellationToken.token,
            ...config,
        });
        const promise = axiosPromise.then((res) => res.data);
        return { promise, cancellationToken, axiosPromise };
    }

    // private xhrPost<T>(
    //     url: string,
    //     data: unknown,
    //     config?: AxiosRequestConfig
    // ) {
    //     const cancellationToken = this.getNewAxiosCancellationToken();
    //     const axiosPromise = this.axios.post<T>(url, data, {
    //         cancelToken: cancellationToken.token,
    //         ...config,
    //     });
    //     const promise = axiosPromise.then((res) => res.data);
    //     return { promise, cancellationToken, axiosPromise };
    // }

    // private xhrDelete<T>(url: string) {
    //     const cancellationToken = this.getNewAxiosCancellationToken();
    //     const axiosPromise = this.axios.delete<T>(url, {
    //         cancelToken: cancellationToken.token,
    //     });
    //     const promise = axiosPromise.then((res) => res.data);
    //     return { promise, cancellationToken, axiosPromise };
    // }

    public lookupUserByEmail(
        email: string | readonly string[],
        config?: AxiosRequestConfig
    ) {
        const value = (typeof email === "string" ? [email] : email).join(",");

        return this.xhrGet<[] | [UserInfo]>(`${ENDPOINTS.USERS}`, {
            params: {
                email: value,
            },
            ...config,
        });
    }
    public lookupUserById(
        id: string | readonly string[],
        config?: AxiosRequestConfig
    ) {
        const value = (typeof id === "string" ? [id] : id).join(",");

        return this.xhrGet<[] | [UserInfo]>(`${ENDPOINTS.USERS}`, {
            params: {
                id: value,
            },
            ...config,
        });
    }
}
