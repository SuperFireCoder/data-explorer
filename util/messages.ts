import { getWorkFlowUrl, IS_DEVELOPMENT } from "./env";

export const sendDatasetId = (dataUuid: string) => {
    const EVENT_NAME = "select-dataset";
    const TARGET_ORIGIN = IS_DEVELOPMENT ? '*' : getWorkFlowUrl();

    if (window.parent){
        try {
            window.parent.postMessage({
                event_id: EVENT_NAME,
                dataset_data: {
                    id:dataUuid
                }
            }, TARGET_ORIGIN);

            console.log("messages.sendDatasetId: datauuid", EVENT_NAME, dataUuid, TARGET_ORIGIN);
        } catch (error){
            console.error(error);
        }
    }
};