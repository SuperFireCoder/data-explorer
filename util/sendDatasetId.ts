import { getWorkFlowUrl, IS_DEVELOPMENT } from "./env";

export const sendDatasetId = (dataUuid: string) => {
    const EVENT_NAME = "select-dataset";

    if (window.parent){
        console.log("sendDatasetId: datauuid", EVENT_NAME, dataUuid);
        window.parent.postMessage({
            event_id: EVENT_NAME,
            dataset_data: {
                id:dataUuid
            }
        }, IS_DEVELOPMENT ? '*' : getWorkFlowUrl());
    }
};