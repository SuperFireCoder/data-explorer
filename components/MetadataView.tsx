import { useEffect, useState } from "react";
import axios from "axios";
import { H4, H6, Icon } from "@blueprintjs/core";

import { useKeycloakInfo } from "../util/keycloak";
import { getDataExplorerBackendServerUrl } from "../util/env";
import { Dataset } from "../interfaces/Dataset";

export interface Props {
    datasetId: string;
}

export default function MetadataView({ datasetId }: Props) {
    const { keycloak } = useKeycloakInfo();
    const keycloakToken = keycloak?.token;

    const [metadata, setMetadata] = useState<
        | { type: "dataset"; data: Dataset }
        | { type: "error"; error: any }
        | undefined
    >(undefined);

    useEffect(
        function loadMetadata() {
            const cancellationToken = axios.CancelToken.source();

            (async () => {
                try {
                    const headers: Record<string, string> = {};

                    if (keycloakToken && keycloakToken.length > 0) {
                        headers["Authorization"] = `Bearer ${keycloakToken}`;
                    }

                    const {
                        data,
                    } = await axios.get(
                        `${getDataExplorerBackendServerUrl()}/api/dataset/${datasetId}`,
                        { headers }
                    );

                    setMetadata({ type: "dataset", data });
                } catch (e) {
                    // Ignore cancellation
                    if (axios.isCancel(e)) {
                        return;
                    }

                    // Set error state
                    setMetadata({ type: "error", error: e });
                }
            })();

            return function stopLoadMetadata() {
                cancellationToken.cancel();
            };
        },
        [keycloakToken, datasetId]
    );

    switch (metadata?.type) {
        case undefined: {
            return <p>Please wait while we fetch this dataset...</p>;
        }

        case "error": {
            return (
                <div>
                    <H4>
                        <Icon icon="error" intent="danger" /> Error
                    </H4>
                    <p>An error occurred when fetching this dataset</p>
                </div>
            );
        }

        case "dataset": {
            const data = metadata.data;
            const bccvlMetadata = data["bccvl:metadata"];

            const description = bccvlMetadata.description;

            const displayedMetadata = {
                // Species
                "Scientific name": bccvlMetadata.scientificName?.join(" "),
                // Raster
                Resolution: bccvlMetadata.resolution,
                // Dataset
                "Year range":
                    bccvlMetadata.year_range &&
                    `${bccvlMetadata.year_range[0]} to ${bccvlMetadata.year_range[1]}`,
                Domain: bccvlMetadata.domain,
                Genre: bccvlMetadata.genre,
                Categories: bccvlMetadata.categories.join(" "),
                // Citation, referencing and licensing
                DOI: bccvlMetadata.doi,
                Attributions: bccvlMetadata.attributions && (
                    <ul>
                        {bccvlMetadata.attributions.map(({ type, value }) => (
                            <li key={type + value}>
                                <H6>{type}</H6>
                                <p>{value}</p>
                            </li>
                        ))}
                    </ul>
                ),
                Acknowledgement: bccvlMetadata.acknowledgement,
                License: bccvlMetadata.license,
                "Landing page": bccvlMetadata.landingpage && (
                    <a href={bccvlMetadata.landingpage} target="_blank">
                        {bccvlMetadata.landingpage}
                    </a>
                ),
            };

            return (
                <div>
                    {description && <p>{description}</p>}
                    <ul>
                        {Object.entries(displayedMetadata).map(
                            ([field, value]) =>
                                value && (
                                    <li key={field}>
                                        <H6>{field}</H6>
                                        <p>{value}</p>
                                    </li>
                                )
                        )}
                    </ul>
                </div>
            );
        }
    }
}
