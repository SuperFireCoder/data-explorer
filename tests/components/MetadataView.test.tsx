import { act, render, screen } from "@testing-library/react";
import MetadataView from "../../components/MetadataView";
import mockAxios from "jest-mock-axios";

import * as keycloakUtil from "../../util/keycloak";

// Mock Keycloak
const defaultKeycloakInfo = {
    keycloak: {
        authenticated: false,
    },
    initialized: false,
};

jest.mock("../../util/keycloak", () => ({
    useKeycloakInfo: jest.fn(() => defaultKeycloakInfo),
}));

// Mock environment utils
jest.mock("../../util/env", () => ({
    getDataExplorerBackendServerUrl: () => "http://test.backend.example.com",
}));

afterEach(() => {
    // Reset mocks
    keycloakUtil.useKeycloakInfo
        .mockReset()
        .mockImplementation(() => defaultKeycloakInfo);
    mockAxios.reset();
});

describe("MetadataView", () => {
    it("starts with a 'please wait' message", () => {
        render(<MetadataView datasetId="test-dataset-id" />);

        expect(
            screen.getByText("Please wait while we fetch this dataset", {
                exact: false,
            })
        ).toBeTruthy();
    });

    it("requests metadata info from backend on mount", () => {
        render(<MetadataView datasetId="test-dataset-id" />);

        expect(mockAxios.get).toBeCalledTimes(1);
        expect(mockAxios.get.mock.calls[0][0]).toBe(
            "http://test.backend.example.com/api/dataset/test-dataset-id"
        );
    });

    it("renders metadata from server", async () => {
        render(<MetadataView datasetId="test-dataset-id" />);

        await act(async () => {
            mockAxios.mockResponse({
                data: {
                    "bccvl:metadata": {
                        scientificName: ["Panthera leo"],
                        resolution: "Really fine",
                        domain: "The mighty jungle",
                        genre: "Mbube",
                        categories: ["biological", "occurrence"],
                        doi: "10.0000/0000000000",
                        attributions: [
                            {
                                type: "Citation",
                                value: "The Lion Sleeps Tonight",
                            },
                        ],
                        license: "Public Domain",
                    },
                    rangeAlternates: {},
                },
            });

            // Arbitrarily wait for all Promises to complete for state change to
            // be applied as a result of axios XHR above
            await Promise.resolve();
        });

        expect(screen.getByTestId("metadata-view")).toMatchSnapshot();
    });

    it("renders error when XHR fails", async () => {
        render(<MetadataView datasetId="test-dataset-id" />);

        await act(async () => {
            mockAxios.mockError(new Error("HTTP 404 Not Found"));

            // Arbitrarily wait for all Promises to complete for state change to
            // be applied as a result of axios XHR above
            await Promise.resolve();
        });

        expect(
            screen.getByText("An error occurred when fetching this dataset", {
                exact: false,
            })
        ).toBeTruthy();
    });

    it("requests metadata info from backend on mount", () => {
        // Mock keycloak
        const keycloakInfo = {
            ...defaultKeycloakInfo,
            keycloak: {
                authenticated: true,
                token: "TEST_KEYCLOAK_TOKEN_USUALLY_BASE64_ENCODED",
            },
        };

        jest.spyOn(keycloakUtil, "useKeycloakInfo").mockImplementation(
            () => keycloakInfo
        );

        render(<MetadataView datasetId="test-dataset-id" />);

        expect(mockAxios.get.mock.calls[0][1]).toHaveProperty(
            ["headers", "Authorization"],
            "Bearer TEST_KEYCLOAK_TOKEN_USUALLY_BASE64_ENCODED"
        );
    });
});
