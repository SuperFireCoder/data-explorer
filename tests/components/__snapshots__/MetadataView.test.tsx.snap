// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataView renders metadata from server 1`] = `
<div
  data-testid="metadata-view"
>
  <ul>
    <li>
      <b>
        Scientific name
      </b>
      <p>
        Panthera leo
      </p>
    </li>
    <li>
      <b>
        Resolution
      </b>
      <p>
        Really fine
      </p>
    </li>
    <li>
      <b>
        Variables
      </b>
      <p>
        test
      </p>
    </li>
    <li>
      <b>
        Domain
      </b>
      <p>
        The mighty jungle
      </p>
    </li>
    <li>
      <b>
        Genre
      </b>
      <p>
        Mbube
      </p>
    </li>
    <li>
      <b>
        Categories
      </b>
      <p>
        biological occurrence
      </p>
    </li>
    <li>
      <b>
        DOI
      </b>
      <p>
        10.0000/0000000000
      </p>
    </li>
    <li>
      <b>
        Attributions
      </b>
      <div>
        <ul>
          <li>
            <b>
              Citation
            </b>
            <p>
              The Lion Sleeps Tonight
            </p>
          </li>
        </ul>
      </div>
    </li>
    <li>
      <b>
        License
      </b>
      <p>
        Public Domain
      </p>
    </li>
  </ul>
</div>
`;
